### DISCUSSION TITLE
Static CTOR Reference for `Func<>` and Delegate

### DISCUSSION THREAD

**Discussion by TheBrambleShark** [dotnet/roslyn#75085] [Ideas] [answered]
When dealing with methods which accept a `Func<>` parameter (or for that matter, a `Func<>` property or member variable), it is possible to pass a reference to a method that matches the same signature.

For instance, the following code is valid:
```cs
var evenNumbers = Enumerable.Range(1, 10).Where(int.IsEvenInteger); // 2, 4, 6, 8, 10
```

This particular example uses a static method, but member methods are also permitted. In this example, the `Where()` call is expanded into (something akin to) `Where(it => it.IsEvenInteger(it))`.

However, there is no similar use case for accessing static CTORs. In a use case I use fairly commonly, I may want to transform one type into another with `Select()`. As an example:

```cs
IEnumerable<DirectoryInfo> childDirectories =
    Directory
    .EnumerateDirectories(@"C:\Temp\Foo")
    .Select(it => new DirectoryInfo(it));
```

In this particular example, I could simply modify the code to be `new DirectoryInfo(@"C:\Temp\Foo).EnumerateDirectories()`, but it serves to show my point.

What I would like to do is to be able to pass a static reference to the ctor, just as I can with `int.IsEvenInteger` in the first example.

### Proposed Syntax

Option 1 is to take inspiration from Kotlin. I propose `Where(Foo::ctor)`, which while is new syntax, should not cause any weird edge cases. This also won't involve any language changes as the compiler would simply transform this into `Where(it => new Foo(it))`, possibly with some efficiencies or shortcuts when lowering.

Option 2 is to take inspiration from static functions. I propse `Where(Foo.Foo)`. This is already reserved (you can't have a static function in `Foo` named `Foo`), but I worry that treating the ctor like a static member function may cause some unexpected behavior, or may lead to incorrect expecations on the part of developers.

I'm definitely open to other syntax suggestions, but ideally I'd like to maximise backwards compatibility. I'd imagine that, if properly designed, this is just syntax fluff over existing functionality and should in theory be compatible with any version of the language that supports `Func<>` or Delegate.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/discussions/674. Also, please make sure that you're opening language suggestions on the csharplang repo, not on roslyn. Thanks!

**Reply by TheBrambleShark** [reply]
Apologies, seems there was some confusion as to the right place for this. I'll keep that in mind for the future. Thanks for the link!
